extensions/amp-fresh/0.1/amp-fresh-manager.js

```diff
@@ -14,14 +14,98 @@
  * limitations under the License.
  */
 
+import {addParamToUrl} from '../../../src/url';
 import {fromClass} from '../../../src/service';
+import {isExperimentOn} from '../../../src/experiments';
+import {whenDocumentReady} from '../../../src/document-ready';
+import {xhrFor} from '../../../src/xhr';
 
 
+/** @const */
+const TAG = 'amp-fresh';
+
 /**
  * Manages registered amp-fresh components and does the xhr request
  * and update.
  */
-export class AmpFreshManager {}
+export class AmpFreshManager {
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+
+    /** @type {!Window} */
+    this.win = win;
+
+    if (!isExperimentOn(this.win, TAG)) {
+      return;
+    }
+
+    /** @private @const {!Object<string, !./amp-fresh.AmpFresh>} */
+    this.ampFreshInstances_ = Object.create(null);
+
+    /** @private {?Document} */
+    this.docPromise_ = null;
+
+    this.fetchDocument_().then(doc => {
+      this.update_(doc);
+    });
+  }
+
+  /**
+   * Registers an amp-fresh component instance to be managed by this windows
+   * amp-fresh-manager instance.
+   * @param {string} id
+   * @param {!./amp-fresh.AmpFresh} ampFreshIntance
+   */
+  register(id, ampFreshInstance) {
+    this.ampFreshInstances_[id] = ampFreshInstance;
+  }
+
+  /**
+   * @return {!Promise<!Document>}
+   */
+  fetchDocument_() {
+    if (this.docPromise_) {
+      return;
+    }
+    const url = addParamToUrl(this.win.location.href, 'amp-fresh', '1');
+    this.docPromise_ = Promise.all([
+      xhrFor(this.win).fetchDocument(url),
+      whenDocumentReady(this.win.document),
+    ]).then(args => args[0]);
+
+    this.docPromise_.catch(() => {
+      this.onFetchDocumentFailure_();
+    });
+    return this.docPromise_;
+  }
+
+  /**
+   * @param {!Document} doc
+   */
+  update_(doc) {
+    const ampFreshFromServerItems = Array.prototype.slice.call(
+        doc.getElementsByTagName('amp-fresh'));
+    ampFreshFromServerItems.forEach(ampFreshFromServer => {
```

Comment:
the xhr code in `fetchDocument_` `xhrFor(this.win).fetchDocument(url)` should be fetching the latest document. By adding the `?amp-fresh=1` query param this *should* (to be implemented) force the backend to guarantee that current document that it responds with in the xhr has been fetched in the last couple of minutes.

the problem again is that even though the cached document is only 1 behind, for long tail pages that 1 version behind could possibly be a year old.